src/tools/perf/lib/libperf.c

```diff
@@ -1160,32 +1170,52 @@ static void ucp_perf_test_free_mem(ucx_perf_context_t *perf)
 static void ucp_perf_test_destroy_eps(ucx_perf_context_t* perf)
 {
     unsigned i, thread_count = perf->params.thread_count;
-    ucs_status_ptr_t    *req;
-    ucs_status_t        status;
+    unsigned num_in_prog     = 0;
+    ucs_status_ptr_t **reqs;
+    ucs_status_t status;
+
+    reqs = ucs_malloc(thread_count * sizeof(*reqs), "ep_close_reqs");
+    if (reqs == NULL) {
+        ucs_warn("failed to allocate memory for UCP requests");
+        return;
+    }
 
     for (i = 0; i < thread_count; ++i) {
         if (perf->ucp.tctx[i].perf.ucp.rkey != NULL) {
             ucp_rkey_destroy(perf->ucp.tctx[i].perf.ucp.rkey);
         }
 
         if (perf->ucp.tctx[i].perf.ucp.ep != NULL) {
-            req = ucp_ep_close_nb(perf->ucp.tctx[i].perf.ucp.ep,
-                                  UCP_EP_CLOSE_MODE_FLUSH);
+            reqs[i] = ucp_ep_close_nb(perf->ucp.tctx[i].perf.ucp.ep,
+                                      UCP_EP_CLOSE_MODE_FLUSH);
 
-            if (UCS_PTR_IS_PTR(req)) {
-                do {
-                    ucp_worker_progress(perf->ucp.tctx[i].perf.ucp.worker);
-                    status = ucp_request_check_status(req);
-                } while (status == UCS_INPROGRESS);
-
-                ucp_request_release(req);
-            } else if (UCS_PTR_STATUS(req) != UCS_OK) {
+            if (UCS_PTR_IS_PTR(reqs[i])) {
+                ++num_in_prog;
+            } else if (UCS_PTR_STATUS(reqs[i]) != UCS_OK) {
                 ucs_warn("failed to close ep %p on thread %d: %s\n",
                          perf->ucp.tctx[i].perf.ucp.ep, i,
-                         ucs_status_string(UCS_PTR_STATUS(req)));
+                         ucs_status_string(UCS_PTR_STATUS(reqs[i])));
+            }
+        }
+    }
+
+    while (num_in_prog != 0) {
+        ucp_perf_worker_progress(perf);
+        for (i = 0; i < thread_count; ++i) {
+            if (!UCS_PTR_IS_PTR(reqs[i])) {
+                continue;
+            }
+
+            status = ucp_request_check_status(reqs[i]);
+            if (status != UCS_INPROGRESS) {
+                --num_in_prog;
+                ucp_request_release(reqs[i]);
+                reqs[i] = NULL;
```

Comment:
maybe remove req from the array (reqs[i]=reqs[--num-in_prog])
also, not add NULL reqs to array after close_nb

Reply:
done